src/diffusers/schedulers/scheduling_consistency_models.py

```diff
@@ -0,0 +1,330 @@
+            set to 7.0 in the original implementation, which is also the original value suggested in the EDM paper.
+        clip_denoised (`bool`):
+            Whether to clip the denoised outputs to `(-1, 1)`. Defaults to `True`.
+        timesteps (`List` or `np.ndarray` or `torch.Tensor`, *optional*):
+            Optionally, an explicit timestep schedule can be specified. The timesteps are expected to be in increasing
+            order.
+    """
+
+    order = 1
+
+    @register_to_config
+    def __init__(
+        self,
+        num_train_timesteps: int = 40,
+        sigma_min: float = 0.002,
+        sigma_max: float = 80.0,
+        sigma_data: float = 0.5,
+        s_noise: float = 1.0,
+        rho: float = 7.0,
+        clip_denoised: bool = True,
+        timesteps: Optional[Union[List, np.ndarray, torch.Tensor]] = None,
+    ):
+        # standard deviation of the initial noise distribution
+        self.init_noise_sigma = sigma_max
+
+        # setable values
+        self.num_inference_steps = None
+        self.timesteps = timesteps
+        self.is_scale_input_called = False
+
+    def index_for_timestep(self, timestep, schedule_timesteps=None):
+        if schedule_timesteps is None:
+            schedule_timesteps = self.timesteps
+
+        indices = (schedule_timesteps == timestep).nonzero()
+        return indices.item()
+
+    def scale_model_input(
+        self, sample: torch.FloatTensor, timestep: Union[float, torch.FloatTensor]
+    ) -> torch.FloatTensor:
+        """
+        Scales the consistency model input by `(sigma**2 + sigma_data**2) ** 0.5`, following the EDM model.
+
+        Args:
+            sample (`torch.FloatTensor`): input sample
+            timestep (`float` or `torch.FloatTensor`): the current timestep in the diffusion chain
+        Returns:
+            `torch.FloatTensor`: scaled input sample
+        """
+        # Get sigma corresponding to timestep
+        if isinstance(timestep, torch.Tensor):
+            timestep = timestep.to(self.timesteps.device)
+        step_idx = self.index_for_timestep(timestep)
+        sigma = self.sigmas[step_idx]
+
+        sample = sample / ((sigma**2 + self.config.sigma_data**2) ** 0.5)
+
+        self.is_scale_input_called = True
+        return sample
+
+    def scale_timestep(self, timestep: Union[float, torch.FloatTensor]):
+        """
+        Scales the timestep based on the associated Karras sigma, for input to the consistency model.
+
+        Args:
+            timestep (`float` or `torch.FloatTensor`): the current timestep in the diffusion chain
+        Returns:
+            `torch.FloatTensor`: scaled input timestep
+        """
+        # Get sigma corresponding to timestep
+        if isinstance(timestep, torch.Tensor):
+            timestep = timestep.to(self.timesteps.device)
+        step_idx = self.index_for_timestep(timestep)
+        sigma = self.sigmas[step_idx]
+
+        return 1000 * 0.25 * torch.log(sigma + 1e-44)
```

Comment:
Is this really scaling the timestep if the output is a function of the sigma? Sorry I'm not super familiar with this part :)

Reply:
In the `denoise` function in the original code, this quantity (`1000 * 0.25 * torch.log(sigma + 1e-44)`) is called `rescaled_t` and input as the timestep argument to the U-Net (https://github.com/openai/consistency_models/blob/main/cm/karras_diffusion.py#L334-L349). I guess although the authors describe a consistency model as being parameterized by a time-conditioned U-Net as $`F_\theta(x, t)`$, in practice $`t`$ is actually a `sigma` (noise scale) rather than a raw `timestep`.

This method is very similar to the `timestep` code in `StableDiffusionLatentUpscalePipeline` linked in https://github.com/huggingface/diffusers/pull/3492#discussion_r1218551132 and seems to be based on the noise pre-conditioning parameter $`c_{\text{noise}}(\sigma)`$ in the EDM column of Table 1 of [Karras et al. 2022](https://arxiv.org/pdf/2206.00364.pdf). So perhaps one alternative is to move the logic into the denoising loop of the pipeline, like it is in `StableDiffusionLatentUpscalePipeline`.